### PR TITLE
add warnings for unsupported floats

### PR DESCRIPTION
**Pull request by asledz**
## Before submitting checklist

- [x] Did you **update the CHANGELOG**? (not for test updates, internal changes/refactors or CI/CD setup)
- [x] Did you **ask the docs owner** to review all the user-facing changes?

### CODE REVIEW

File: src/neptune/attributes/atoms/float.py
Comment (Raalsky) [inline src/neptune/attributes/atoms/float.py:55]: Remember to ask @normandy7 to review the warning content.

File: src/neptune/types/series/float_series.py
Comment (Raalsky) [inline src/neptune/types/series/float_series.py:108]: list comprehension + filter is probably faster. Use `append` in Python only in rare scenarios.

File: tests/unit/neptune/new/test_stringify_unsupported.py
Comment (Raalsky) [inline tests/unit/neptune/new/test_stringify_unsupported.py:537]: Let's test it somewhere else, I don't think it's that related to stringify value

File: src/neptune/utils.py
Comment (Raalsky) [inline src/neptune/utils.py:17]: That's a module for user-exposed utils. This function is much more internal.

File: tests/unit/neptune/new/test_stringify_unsupported.py
Comment (Raalsky) [inline tests/unit/neptune/new/test_stringify_unsupported.py:540]: What will happen here? Do we log an operation to assign value or like in series we're going to not log a unsupported value?

File: src/neptune/types/series/float_series.py
Comment (asledz) [reply src/neptune/types/series/float_series.py]: Right! Changed that :D

File: tests/unit/neptune/new/test_stringify_unsupported.py
Comment (asledz) [reply tests/unit/neptune/new/test_stringify_unsupported.py]: Moved the tests 😇

File: src/neptune/utils.py
Comment (asledz) [reply src/neptune/utils.py]: True, deleted from here 😄

File: tests/unit/neptune/new/test_stringify_unsupported.py
Comment (asledz) [reply tests/unit/neptune/new/test_stringify_unsupported.py]: Like in series:
- throw the warning to user, that is not a supported value
- skip the value

File: src/neptune/types/series/float_series.py
Comment (asledz) [reply src/neptune/types/series/float_series.py]: Done

File: src/neptune/attributes/atoms/float.py
Comment (asledz) [reply src/neptune/attributes/atoms/float.py:55]: It is copied from previous PR, which she reviewed 😄 but asked her to doublecheck, that everything is correct

File: src/neptune/attributes/atoms/float.py
Comment (normandy7) [inline src/neptune/attributes/atoms/float.py:58]:
```suggestion
                f"For now, you can use utils.stringify_unsupported() to log one or more values as strings: "
```

**Review by normandy7 — DISMISSED** [review]

File: CHANGELOG.md
Comment (Raalsky) [inline CHANGELOG.md:24]: Skip unsupported float values?

Comment (asledz) [reply CHANGELOG.md]: True, that is missing -> changed to skip and warn (I think it is important, because it is not a silent skip)

File: tests/unit/neptune/new/attributes/series/test_float_series.py
Comment (Raalsky) [inline tests/unit/neptune/new/attributes/series/test_float_series.py:62]: Let's test multiple values to the same namespaces:
```python
>>> run['f'].extend([1.5, 2.3, str(float("nan")), 4.7])
```
or:
```python
>>> run['b'].extend([str(float("inf")), 5.5, 2.3, 6.7])
```

File: tests/unit/neptune/new/attributes/atoms/test_float.py
Comment (Raalsky) [inline tests/unit/neptune/new/attributes/atoms/test_float.py:74]: Can we assert here that these particular namespaces weren't created?

File: src/neptune/utils.py
Comment (Raalsky) [inline src/neptune/utils.py:17]: Let's extend this list with your function

File: src/neptune/types/series/float_series.py
Comment (Raalsky) [inline src/neptune/types/series/float_series.py:108]: How about `steps` and `timestamps` here? As we're getting rid of particular value we have to delete corresponding timestamp and step as well

File: tests/unit/neptune/new/attributes/series/test_float_series.py
Comment (asledz) [reply tests/unit/neptune/new/attributes/series/test_float_series.py:62]: Added 😄

File: tests/unit/neptune/new/attributes/atoms/test_float.py
Comment (asledz) [reply tests/unit/neptune/new/attributes/atoms/test_float.py:78]: Added! 😄

File: src/neptune/utils.py
Comment (asledz) [reply src/neptune/utils.py]: Extended back 😅

File: src/neptune/types/series/float_series.py
Comment (asledz) [reply src/neptune/types/series/float_series.py]: True, missed that! both of those lists are now cleared 😄

File: tests/unit/neptune/new/attributes/atoms/test_float.py
Comment (Raalsky) [reply tests/unit/neptune/new/attributes/atoms/test_float.py:78]: `run.exists("infinity")`? `__dict__` is internal property and I'm not sure that our abstractions handle it properly

File: src/neptune/utils.py
Comment (Raalsky) [reply src/neptune/utils.py:17]: Looks like it's still here :D.

Comment (asledz) [reply src/neptune/utils.py]: Moved back because of the comment you left later on, but decided to move to internal common module 😉

File: src/neptune/types/series/float_series.py
Comment (Raalsky) [inline src/neptune/types/series/float_series.py:107]: How about something more generic like:
```python
def filter_unsupported_values(values, steps, timestamps, filter_by):
    filtered = [
        (value, step, timestamp) for value, step, timestamp in zip(values, steps, timestamps) if filter_by(value)
    ]
    return (
        [value for value, _, _ in filtered],
        [step for _, step, _ in filtered],
        [timestamp for _, _, timestamp in filtered]
    )
```
and usage:
```python
        if steps is None:
            filled_steps = list(islice(cycle([None]), len(values)))
        else:
            assert len(values) == len(steps)
            filled_steps = steps

        if timestamps is None:
            filled_timestamps = list(islice(cycle([time.time()]), len(values)))
        else:
            assert len(values) == len(timestamps)
            filled_timestamps = timestamps

        clean_values, self._steps, self._timestamps = filter_unsupported_values(
            values=values,
            steps=filled_steps,
            timestamps=filled_timestamps,
            filter_by=is_unsupported_float
        )
        self._values = [float(value) for value in clean_values]
```

But it wasn't tested :D.

File: src/neptune/utils.py
Comment (Raalsky) [reply src/neptune/utils.py]: Sorry for confusing you, one thing is the proper placing another is to remember about `__all__` filling up.

File: src/neptune/types/series/float_series.py
Comment (asledz) [reply src/neptune/types/series/float_series.py]: That looks pretty! But we still need to warn user about unsupported, please take a look if my solution for that looks as pretty as this 😇

File: src/neptune/attributes/atoms/float.py
Comment (Raalsky) [inline src/neptune/attributes/atoms/float.py:58]: That's not true :wink:. `stringify_unsupported` doesn't help here or I don't see how it should work with float series.

Comment (asledz) [reply src/neptune/attributes/atoms/float.py:58]: I mean, isn't it? it is applicable for atom floats. What might be a bit misleading maybe is "one or more values as strings" part, cause it is not useful in float_series. 
Is this is the part you say is not true? 😅 
(I am not the best with warning messages 😅 )

File: src/neptune/types/series/float_series.py
Comment (asledz) [reply src/neptune/types/series/float_series.py]: Yep, fixed with the generic result

File: src/neptune/attributes/atoms/float.py
Comment (normandy7) [reply src/neptune/attributes/atoms/float.py:58]: Since I contributed to the message: My assumption was that this doesn't show for float series, but rather if the user tries to log a single value (or collection of values, hence "one or more"), in which case `stringify_unsupported` could be used.

Comment (Raalsky) [reply src/neptune/attributes/atoms/float.py:58]: Keep in mind that `stringify_unsupported` means that if there will be some unsupported value (based on context as it's lazy-evaluated) it's going to be cased to a String. String may not be proper type in all of the context like -> trying to append a String value to FloatSeries.

To summarize about `stringify_unsupported`:
- It should work with atoms in most of the cases as `stringify_unsupported` is going to be either a proper value or a String; but in case there was a float value previously or something similar it may raise a MetadataInconsistency/ValueError as we're trying to do something similar to:
```python
>>> run['a'] = 4.2
>>> run['a'] = 'unsupported'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/neptune/Repos/neptune-client/src/neptune/metadata_containers/metadata_container.py", line 105, in inner_fun
    return fun(self, *args, **kwargs)
  File "/Users/neptune/Repos/neptune-client/src/neptune/metadata_containers/metadata_container.py", line 312, in __setitem__
    self.__getitem__(key).assign(value)
  File "/Users/neptune/Repos/neptune-client/src/neptune/handler.py", line 88, in inner_fun
    return fun(self, *args, **kwargs)
  File "/Users/neptune/Repos/neptune-client/src/neptune/handler.py", line 222, in assign
    attr.process_assignment(value, wait=wait)
  File "/Users/neptune/Repos/neptune-client/src/neptune/attributes/attribute.py", line 66, in process_assignment
    return self.assign(value, wait=wait)
  File "/Users/neptune/Repos/neptune-client/src/neptune/attributes/atoms/float.py", line 51, in assign
    value = FloatVal(value)
  File "/Users/neptune/Repos/neptune-client/src/neptune/types/atoms/float.py", line 39, in __init__
    self.value = float(extract_if_stringify_value(value))
ValueError: could not convert string to float: 'unsupported'
```
but I don't have a good example yet.

- It shouldn't work for float/image series as `.extend(stringify_unsupported([5.4, 6.7, float(inf), 7.8]))` in the base case will lead to something like `.extend([5.4, 6.7, "inf", 7.8])` so it's the same as trying to append a String to a FloatSeries (should lead to an exception like MetadataInconsistency)

Comment (Raalsky) [reply src/neptune/attributes/atoms/float.py:58]: Btw. Another thing that I spotted is that we're going to show double warning message with:
```python
>>> run['a'] = stringify_unsupported(float('inf'))
The value you're trying to log is an unsupported float value (inf) and will be logged as string. We'll add support for these types of values in the future.
<stdin>:1: NeptuneUnsupportedValue: WARNING: The value you're trying to log is a nonstandard float value (inf) that is not currently supported. We'll add support for this type of value in the future. For now, you can use utils.stringify_unsupported() to log one or more values as strings: run['field'] = stringify_unsupported(float(inf)
```
Is it expected?

File: src/neptune/common/utils.py
Comment (Raalsky) [inline src/neptune/common/utils.py:273]: `src/neptune/common` means that it's a part of both `neptune.legacy` and regular `neptune` packages. I don't think there are any usage of it in legacy client. Let's move it to something like `src/neptune/internal/types/utils.py`

File: src/neptune/types/series/float_series.py
Comment (Raalsky) [inline src/neptune/types/series/float_series.py:116]: You can hide it under the `filter_by` implementation -> Instead of calling `is_unsupported_float` directly you may wrap it with something like:
```python
def is_unsupported_float_with_warn(x):
    if is_unsupported_float(x):
        warn_once(...)
        return ....
    return ...
```

Comment (asledz) [reply src/neptune/types/series/float_series.py:115]: Done!

File: src/neptune/common/utils.py
Comment (asledz) [reply src/neptune/common/utils.py]: Moved

File: src/neptune/attributes/atoms/float.py
Comment (asledz) [reply src/neptune/attributes/atoms/float.py:58]: About the double warning -> 
true, not expected. But I was thinking, using stringify_unsupported should warn us once, so I deleted stringify_unsupported warning, now only the NeptuneUnsupportedValue will be displayed. 
I think it is a good solution for end user, @normandy7 and @Raalsky what do you think? 🤔

File: src/neptune/attributes/atoms/float.py
Comment (normandy7) [inline src/neptune/attributes/atoms/float.py:59]: Add missing closing parenthesis

```suggestion
                f"run['field'] = stringify_unsupported(float({str(value.value)}))",
```

File: CHANGELOG.md
Comment (Raalsky) [inline CHANGELOG.md:23]: Could you remove this entry as well?

Comment (asledz) [reply CHANGELOG.md]: Removed!

File: src/neptune/attributes/atoms/float.py
Comment (asledz) [reply src/neptune/attributes/atoms/float.py]: Added

File: src/neptune/attributes/atoms/float.py
Comment (asledz) [reply src/neptune/attributes/atoms/float.py:58]: So now we would have:
Scenario 1: (single float, single warning)
 `run['a'] = float("inf")`
Result 1:
`NeptuneUnsupportedValue: WARNING: The value you're trying to log is a nonstandard float value (inf) that is not currently supported. We'll add support for this type of value in the future. For now, you can use utils.stringify_unsupported() to log one or more values as strings: run['field'] = stringify_unsupported(float(inf)
  run['a'] = float("inf")`
  
Scenario 2: (float in list, warning for float in list, logging correct float)
`run['f'].extend([float("inf"), 4.0])`
Result 2:
`NeptuneUnsupportedValue: WARNING: A value you're trying to log (`inf`) will be skipped because it's a non-standard float value that is not currently supported.
  run['f'].extend([float("inf"), 4.0])`
  
 Scenario 3 (stringify unsupported):
 `run['a'] = stringify_unsupported(float("inf"))`
 Result 3:
 No warning, logged as string -> and I think it makes sense, since this is what user wanted to do in the first place.

File: src/neptune/types/series/float_series.py
Comment (Raalsky) [inline src/neptune/types/series/float_series.py:75]: For not it's just to think of: Do you think we can refactor it someone to not to create an empty FloatSeries whenever someone will but just a single unsupported value?
I mean the following code produces an empty FloatSeries:
```python
>>> run['c'].append(float('nan'))
<stdin>:1: NeptuneUnsupportedValue: WARNING: A value you're trying to log (`nan`) will be skipped because it's a non-standard float value that is not currently supported.
>>> run.get_structure()['c']
<neptune.attributes.series.float_series.FloatSeries object at 0x13ff83d90>
```

Comment (asledz) [reply src/neptune/types/series/float_series.py:75]: I think it requires a bit more rework, I would leave it for a separate pr, after trying to make that change and not looking so obvious 🤔

File: src/neptune/attributes/atoms/float.py
Comment (Raalsky) [reply src/neptune/attributes/atoms/float.py:58]: I rather agree with the current behavior :wink:

File: src/neptune/types/series/float_series.py
Comment (Raalsky) [reply src/neptune/types/series/float_series.py:75]: Ok. Posted to Jira then

**Review by Raalsky — APPROVED** [review]

File: src/neptune/attributes/atoms/float.py
Comment (normandy7) [reply src/neptune/attributes/atoms/float.py:58]: For the record, I also agreed with the behavior 😄

Comment (Raalsky) [reply src/neptune/attributes/atoms/float.py:58]: > For the record

@asledz Just an FYI the general process is to wait for a double approval from some dev and @normandy7 if was involved/asked for review.

Comment (asledz) [reply src/neptune/attributes/atoms/float.py:58]: Oh sorry! Will wait in future, I just assumed a 👍🏻 emoji meant I am free to merging! 😅 Won't do that again 😇

**Pull request merged**
